### ISSUE TITLE
[feature request] cli add possibility to replace annotations

### ISSUE THREAD

**Issue by lublak** [pdfcpu/pdfcpu#435]
Currently you can list or remove annotations.
It would be also usefull to replace them. 
Like `pdfcpu annotations replaceto link|page inFile objNr url|pagenumber`

**Comment by hhrutter** [comment]
Adding and even updating annotations is a tricky thing when you want to process PDFs that were written by any PDF writer out there. I put it in the pipeline.

Thank you for using pdfcpu 💚

**Comment by anythingwithawire** [comment]
yeah, add annotations is the single standout feature that would take this from "why not just use pdftk" to becoming the goto pdf utility available.

(plus export/import named destionations and/or same for bookmarks with convert bookmarks to named destinations)

However, for both the benefit of the authors and for anyone else out there that might like to use annotations to make their pdfs have active content, there is a feature many know nothing about called "pdfmark" that means you can still do this, but you have to run the pdf through the distiller (ghostwriter or adobe) again.

See this link for a good example

- https://leancrew.com/all-this/2021/12/automating-the-annotation-of-pdfs/

and the reference manual

- https://opensource.adobe.com/dc-acrobat-sdk-docs/library/pdfmark/index.html

**Comment by hhrutter** [comment]
Always ask yourself if adding annotations is really what you need.

* https://leancrew.com/all-this/2021/12/automating-the-annotation-of-pdfs/ is just adding text to a page.

That's a one liner for pdfcpu's cli.

See also https://pdfcpu.io/core/stamp

**Comment by anythingwithawire** [comment]
> Always ask yourself if adding annotations is really what you need.
> 
> * https://leancrew.com/all-this/2021/12/automating-the-annotation-of-pdfs/ is just adding text to a page.
> 
> That's a one liner for pdfcpu's cli.
> 
> See also https://pdfcpu.io/core/stamp
You totally miss the point, that was just a trivial example of how to use a pdfmark to add content and functionality to an existing pdf.

I a mnot sure you know in full detail what an annotatio nis and what it might do for you.
 
So, yes, I need annotations. I am not just adding text to a page, not by a long shot.

Stamps have limitations - stamp files do not execute JavaScript like a normal PDF file.  For example, there are no document or page events.  Variables and functions cannot be defined in a document script.   The only scripts guaranteed to execute are the calculation scripts of the fields placed on the stamp.  Everything necessary to execute this script must be present in the script.

There are some (quite arkward) ways to get around quite a few of stamps limitations, but for some of the more serious add ons for live pdf functionality, annotations are where it is at without a doubt

Annotations allow a rich setof possibilities to make your pdfs come alive, as of PDF 1.3, the following annotation types are supported:

**Value of subtype key | Description**
Circle                                 | Circle annotation
FileAttachment               | File attachment annotation
FreeText                           | Free text annotation
Highlight                         | Highlight annotation
Ink                                    | Ink annotation
Line                                  | Line annotation
Link                                  | Link annotation
Movie                              | Movie annotation
Popup                             | Pop-up annotation
Sound                             | Sound annotation
Square                            | Square annotation
Stamp                             | Rubber stamp annotation
StrikeOut                       | Strikeout annotation
Text                                 | Text annotation (note)
TrapNet                          | Trap network annotation
Underline                      | Underline annotation
Widget                           | Widget annotation

Each type has its own set of key-value pairs that can be specified, as described in the PDF Reference. Future versions of PDF may introduce new types.
In addition to these types, annotations with unrecognized Subtype values, called custom annotations, are supported. Custom annotations can contain, in addition to the Rect and Subtype keys, arbitrary key-value pairs.

And there is a lot more than that, but it is all in PDF Reference.

**Comment by hhrutter** [comment]
Everybody has their reasons and their own specific usecases.
This must be and is respected.
So point taken.

Annotations are well known (and this is not the place to list them) and have been dealt with by pdfcpu for some time step by step as requested by the community:

```
Go-> pdfcpu help annot
usage: pdfcpu annotations list   [-p(ages) selectedPages] inFile
       pdfcpu annotations remove [-p(ages) selectedPages] inFile [objNr...]
```

There is a reason there are not many free tools around supporting full fledged adding of annotations.
It's just a ton of work.

**Comment by anythingwithawire** [comment]
yeah which is why I shared the pdfmark route - this is by far the easiest
way
